Changes to RTT estimator

janaiyengar opened the pull request:
Makes RTT estimator use ack_delay correctly, and also fixes use of RTT for non-ack-eliciting packets.

kazuho commented on lib/quicly.c:
```diff
@@ -3324,25 +3321,21 @@ static int handle_ack_frame(quicly_conn_t *conn, size_t epoch, quicly_ack_frame_
-     * quicly_loss_on_packet_acked is NOT OnPacketAcked */
+    /* OnPacketAcked and OnPacketAckedCC */
     if (bytes_acked > 0) {
+        quicly_loss_on_newly_acked(&conn->egress.loss);
```
Shouldn't we call `quicly_loss_on_newly_acked` when we receive an ACK for any new PN?

Assuming that is the case, the condition for calling this particular function should be `largest_newly_acked.packet_number != UINT64_MAX`. `bytes_acked > 0` is a test that checks if any ACK-eliciting packet is newly acknowledged before it was deemed lost. I think that is fine for the other calls that we make in this block.

FWIW, I think there might be an error in the [pseudo-code of recovery spec](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#rfc.appendix.B.5). `OnPacketAckedCC` substracts `acked_packet.size` from the number of bytes inflight. I believe that should not be done if the packet has already declared lost.

janaiyengar replied:
On the code: Yes, good point, and it was an oversight. I've changed it now.
On the pseudo-code: Good catch! I'll file an issue.

janaiyengar commented:
@kazuho: I've fixed the RTT estimator a bit more, and I've incorporated the RTT sampling scheme I proposed in https://github.com/quicwg/base-drafts/issues/2568. I've also simplified the loss detection interface a bit. PTAL.

kazuho commented on include/quicly/loss.h:
```diff
@@ -188,18 +188,27 @@ inline void quicly_loss_update_alarm(quicly_loss_t *r, int64_t now, int64_t last
+
+    /* Only use RTT samples for new largest acked */
+    if (r->largest_acked_packet >= largest_newly_acked)
+        return;
```
I am not sure if this escape condition is correct.

When there is no newly acked packet, `largest_newly_acked` is set to UINT64_MAX. In such case, the `if` would always fail (unless you set `r->largest_acked_packet` to UINT64_MAX, which you should not, because we'd never be emitting a PN that's so large).

janaiyengar replied:
Good catch! I've fixed this now.

kazuho commented on lib/quicly.c:
```diff
@@ -3321,27 +3317,11 @@ static int handle_ack_frame(quicly_conn_t *conn, size_t epoch, quicly_ack_frame_
 
-    /* OnPacketAcked */
-    uint32_t latest_rtt = UINT32_MAX, ack_delay = 0;
-    if (largest_newly_acked.packet_number == frame->largest_acknowledged) {
```
Thank you for the changes.

This might be in contrary to what we have discussed, but I think we should not remove this condition.

Consider the case where an endpoint receives ACK(PN=X), then ACK(PN=X-1, X). We cannot calculate the RTT from the second ACK, because we cannot infer the time difference between when the peer received X-1 and X.

It now seems to me that this condition is therefore required.

janaiyengar replied:
Thanks, Kazuho, but I'm not sure I follow. In the case you describe, we want to calculate the RTT from the first or the second ACK but not both, since the RTT is still measured for the largest_acked, which is X in both ACKs. As I now have it, the code will only use the first ACK for the RTT sample but not the second. FWIW, in your example, the predicate above will fail on the second ACK in the old code, and the second ACK will not yield a valid sample in the new code.

janaiyengar replied:
I've added a comment to explain the semantics of this call.

kazuho commented:
Thank you for all the efforts!
